### PR TITLE
feat(COGS): Enable Subscriptions to save and execute with `tenant_ids`

### PR DESCRIPTION
**Pull request by rahul-kumar-saini**
### Overview
- Builds upon #4775
- Introduces the ability to actually save Tenant IDs into Redis and use them when executing the subscription queries

### How does it work?
- Enabling the `save_subscription_with_tenant_ids` will make the `SubscriptionData.to_dict()` method to start returning `tenant_ids` in the dict representation of the subscription data, if my understanding is correct this will:
    - Update the message posted to scheduled-subscriptions to also include a `tenant_ids` field
    - Update the data stored in Redis to include the same
    - Update the request made by `subscriptions_executor` to start including a `tenant_ids` field
    - Update the original request in messages produced to the subscription result topic to include `tenant_ids`
- All in all, it _should_ just work, ie it should just add Tenant IDs and have them accessible at query process time for us to analyze subscriptions queries by organization in querylog

### Gotchas
- Enabling this new form of subscription will require some sort of backfill of existing subscriptions to now include the new Tenant IDs so the stored subscriptions do not differ pre/post change, and this will likely need to be done every time a new Tenant ID (eg `use_case_id`) is introduced and is required for subscriptions query analysis

### CODE REVIEW

**Comment by fpacifici** [comment]
Two general comments.

I would change the name TenantId to something else. TenantId has a specific meaning in our infra. It is the tenant in Single Tenant. Using that name will be quite confusing.

> Enabling this new form of subscription will require some sort of backfill of existing subscriptions to now include the new Tenant IDs so the stored subscriptions do not differ pre/post change, and this will likely need to be done every time a new Tenant ID (eg use_case_id) is introduced and is required for subscriptions query analysis

Why does this need to be done again when a new use_case_id is added ?

File: snuba/subscriptions/data.py [hunk 1]
```diff
@@ -163,13 +172,16 @@ def build_request(
                 custom_processing.append(validator.validate)
         custom_processing.append(partial(self.add_conditions, timestamp, offset))
 
+        tenant_ids = {**self.tenant_ids}
+        tenant_ids["referrer"] = referrer
+        if "organization_id" not in tenant_ids:
+            # TODO: Subscriptions queries should have an org ID
+            tenant_ids["organization_id"] = 1
```
Comment (fpacifici) [inline snuba/subscriptions/data.py:179]: Is this related with this PR or is it an unrelated bug ?

**Comment by rahul-kumar-saini** [comment]
> I am not sure I understand the relationship between tenantId and the use case id for generic metrics subscriptions.

It would be very small code change to enforce `use_case_id` requirement on generic metrics queries this way, the other way to do it would be to introduce a top level field in the [request](https://github.com/getsentry/snuba-sdk/blob/main/snuba_sdk/request.py#L44-L50) itself and enforce that way.

@volokluev what do you think about `use_case_id` as a Tenant ID but not necessarily used to allocate resources?


> does it mean the capacity management is applied only per use case and not per organization id on the generic metrics platform?

It won't be automatically used by capacity management to start allocating resources, we'd have to create some policy for it (which we don't plan on doing as of now).

> Why does this need to be done again when a new use_case_id is added ?

Unless new use case IDs are introduced that divide and replace a previous one, we won't need to update all subscriptions. It needs to be done again if/when a new Tenant ID itself is added, requiring updating all previously saved subscriptions to attribute executed subscriptions queries appropriately.

In general, we're missing org id from subscriptions entirely at the moment. This solution solves that problem along with the `use_case_id` problem which is why I'm inclined to do it this way.

File: snuba/subscriptions/data.py (hunk 1 above)
Comment (rahul-kumar-saini) [reply snuba/subscriptions/data.py:179]: We have no Tenant IDs currently for any subscription query, so for now it's just been Org ID 1 but if we actually wanted to get an insight into which orgs are using how much, we need the Tenant IDs to be propagated through from Sentry instead of just hardcoding values here.

This change is just cleaning this up to now accept org id if it exists (since after [this](https://github.com/getsentry/sentry/pull/57101) and [this](https://github.com/getsentry/snuba/pull/4775) we will start accepting Tenant IDs)

**Comment by volokluev** [comment]
> @volokluev what do you think about use_case_id as a Tenant ID but not necessarily used to allocate resources?

I think it's a fine thing to do. The policy for `use_case_id` can just be that we are tracking its usage. We can decide later if we want to enforce anything

**Comment by rahul-kumar-saini** [comment]
Dig this PR back up if/when one day Subscriptions need to be COGS'd 🫡
